src/AWSCredentials.jl

```diff
@@ -56,392 +60,497 @@ mutable struct AWSCredentials
     user_arn::String
     account_number::String
     expiry::DateTime
-    renew::Union{Function, Nothing}
+    renew::Union{Function, Nothing}  # Function which can be used to refresh credentials
+
+    function AWSCredentials(
+        access_key_id,
+        secret_key,
+        token="",
+        user_arn="",
+        account_number="";
+        expiry=typemax(DateTime),
+        renew=nothing,
+    )
+        return new(access_key_id, secret_key, token, user_arn, account_number, expiry, renew)
+    end
+end
+
+
+# function Base.show(io::IO,c::AWSCredentials)
+#     println(io, string(c.user_arn,
+#                        isempty(c.user_arn) ? "" : " ",
+#                        "(",
+#                        c.account_number,
+#                        isempty(c.account_number) ? "" : ", ",
+#                        c.access_key_id,
+#                        isempty(c.secret_key) ? "" : ", $(c.secret_key[1:3])...",
+#                        isempty(c.token) ? "" : ", $(c.token[1:3])..."),
+#                        c.expiry,
+#                        ")")
+# end
 
-    function AWSCredentials(access_key_id,secret_key,
-                            token="", user_arn="", account_number="";
-                            expiry=typemax(DateTime),
-                            renew=nothing)
-        new(access_key_id, secret_key, token, user_arn, account_number, expiry, renew)
+
+function Base.copyto!(dest::AWSCredentials, src::AWSCredentials)
+    for f in fieldnames(typeof(dest))
+        setfield!(dest, f, getfield(src, f))
     end
 end
 
+
+dot_aws_config_file() = get(ENV, "AWS_CONFIG_FILE", joinpath(homedir(), ".aws", "config"))
+dot_aws_credentials_file() = get(ENV, "AWS_SHARED_CREDENTIALS_FILE", joinpath(homedir(), ".aws", "credentials"))
+localhost_maybe_ec2() = localhost_is_ec2() || isfile("/sys/devices/virtual/dmi/id/product_uuid")
+localhost_is_lambda() = haskey(ENV, "LAMBDA_TASK_ROOT")
+_aws_get_profile() = get(ENV, "AWS_DEFAULT_PROFILE", get(ENV, "AWS_PROFILE", "default"))
+
+
+"""
+    AWSCredentials(profile)
+
+Create an AWSCredentials object, given a provided profile (if not provided 'default' will be
+used).
+
+Checks credential locations in the order:
+    1. Environment Variables
+    2. ~/.aws/credentials
+    3. ~/.aws/config
+    4. EC2 or ECS metadata
+
+# Keywords
+- `profile::AbstractString`: Specific profile used to search for AWSCredentials
+
+# Returns
+- `AWSCredentials`: Default AWS Credentials or ones for the specified profile
+"""
 function AWSCredentials(;profile=nothing)
     creds = nothing
-    renew = Nothing
+    credential_function = nothing
```

Comment:
You may want to have `credential_function` always reference a function. That may make logic elsewhere easier
```suggestion
    credential_function = () -> nothing
```